python/src/specialize.cc

```
@@ -0,0 +1,884 @@
+  if (native_impl_available) {
+    Py_DECREF(data_ptr_method);
+    Py_DECREF(data_ptr_result);
+  } else {
+    Py_DECREF(fallback_method);
+    Py_DECREF(kwargs);
+    Py_DECREF(args_tuple);
+  }
+
+  return {py::handle(type_str), key};
+
+cleanup:
+  Py_XDECREF(dtype);
+  Py_XDECREF(native_spec_obj);
+  Py_XDECREF(data_ptr_method);
+  Py_XDECREF(data_ptr_result);
+  Py_XDECREF(fallback_method);
+  Py_XDECREF(kwargs);
+  Py_XDECREF(args_tuple);
+  return {py::handle(), py::handle()};
+}
+
+static std::pair<py::handle, py::handle>
+handle_bool_type(PyObject *backend, PyObject *arg, bool is_const,
+                 bool specialize_value, bool align) {
+  return {py::handle(u1_str), py::handle()};
+}
+
+static std::pair<py::handle, py::handle>
+handle_float_type(PyObject *backend, PyObject *arg, bool is_const,
+                  bool specialize_value, bool align) {
+  return {py::handle(fp32_str), py::handle()};
+}
+
+static std::pair<py::handle, py::handle>
+handle_tensor_descriptor(PyObject *backend, PyObject *arg, bool is_const,
+                         bool specialize_value, bool align) {
+  return specialize_tensordesc(arg, false);
+}
+
+static std::pair<py::handle, py::handle>
+handle_gluon_tensor_descriptor(PyObject *backend, PyObject *arg, bool is_const,
+                               bool specialize_value, bool align) {
+  return specialize_tensordesc(arg, true);
+}
+
+static std::pair<py::handle, py::handle>
+handle_constexpr_type(PyObject *backend, PyObject *arg, bool is_const,
+                      bool specialize_value, bool align) {
+  return {py::handle(constexpr_str), py::handle(arg)};
+}
+
+static std::pair<py::handle, py::handle>
+handle_jit_function(PyObject *backend, PyObject *arg, bool is_const,
+                    bool specialize_value, bool align) {
+  PyObject *cache_key = PyObject_GetAttr(arg, cache_key_attr);
+  if (!cache_key) {
+    PyErr_Clear();
+    return {py::handle(constexpr_str), py::handle()};
+  }
```

We should raise the error here, not just ignore the cache key.

---

Indeed, the `constexpr_str` shouldn't be here. If this returns a pair of empty handles, the calling function correctly throws the error. The `PyErr_Clear` is debatable. With it, the calling code later sets the error string "failed to specialize argument". Without it, the AttributeError is forwarded as far as my understanding goes.